src/System.Linq/tests/ConcatTests.cs

```diff
@@ -89,142 +70,138 @@ public void SecondNull()
             Assert.Throws<ArgumentNullException>("second", () => Enumerable.Range(0, 0).Concat(null));
         }
 
-        [Fact]
-        public void TwoEnumerableSources()
+        [Theory]
+        [MemberData(nameof(ArraySourcesData))]
+        [MemberData(nameof(ArraySelectSourcesData))]
+        [MemberData(nameof(EnumerableSourcesData))]
+        [MemberData(nameof(NonCollectionSourcesData))]
+        [MemberData(nameof(ListSourcesData))]
+        [MemberData(nameof(ConcatOfConcatsData))]
+        [MemberData(nameof(ConcatWithSelfData))]
+        [MemberData(nameof(ChainedCollectionConcatData))]
+        public void VerifyEquals(IEnumerable<int> expected, IEnumerable<int> actual)
         {
-            VerifyEquals(
-                Enumerable.Range(0, 7),
-                Enumerable.Range(0, 3).Concat(Enumerable.Range(3, 4)));
+            // workaround: xUnit type inference doesn't work if the input type is not T (like IEnumerable<T>)
+            VerifyEqualsWorker(expected, actual);
         }
 
-        [Fact]
-        public void TwoArraySources()
+        private static void VerifyEqualsWorker<T>(IEnumerable<T> expected, IEnumerable<T> actual)
         {
-            VerifyEquals(
-                Enumerable.Range(0, 7),
-                Enumerable.Range(0, 3).ToArray().Concat(Enumerable.Range(3, 4).ToArray()));
+            // All of these transforms should take an enumerable and produce
+            // another enumerable with the same contents.
+            var transforms = new List<EnumerableTransform<T>>
+            {
+                e => e,
+                e => e.ToArray(),
+                e => e.ToList(),
+                e => e.Select(i => i),
+                e => e.Select(i => i).ToArray(),
+                e => e.Where(i => true),
+                e => e.Concat(Array.Empty<T>()),
+                e => ForceNotCollection(e),
+                e => e.Concat(ForceNotCollection(Array.Empty<T>()))
+            };
+            
+            // We run the transforms N^2 times, by testing all transforms
+            // of expected against all transforms of actual.
+            foreach (var outTransform in transforms)
+            {
+                foreach (var inTransform in transforms)
+                {
+                    IEnumerable<T> sameExpected = outTransform(expected);
+                    IEnumerable<T> sameActual = inTransform(actual);
+
+                    Assert.Equal(sameExpected, sameActual);
+                    Assert.Equal(sameExpected.Count(), sameActual.Count());
```

Comment:
You're correct-- I only included this assert because the previous version of `VerifyEquals` had it, but I see now that it's redundant. I'll remove this.